### ISSUE TITLE
Multi-root workspace `:` variables are not supported

### ISSUE THREAD

**Issue by Boon-in-Oz** [microsoft/pylance-release#4405]
## Environment data

-   Language Server version: v2023.5.30
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.7

## Code Snippet

This is a trimmed version of my techart.code-workspace, which lives in game_p9/bin/tools:
```
{
	"folders": [
		{
			"name": "Block Transfer",
			"path": "BlockTrans"
		},
		{
			"name": "Techart Packages",
			"path": "../../../game/bin/teams/techart/packages"
		},
		{
			"name": "Maya",
			"path": "Apps/maya"
		},
		{
			"name": "TechTools",
			"path": "TechTools"
		},
	],
	"settings": {
		"python.defaultInterpreterPath": "${workspaceFolder:TechTools}/../../../../game/bin/teams/techart/python3.7.7/python.exe",
		},
		"python.analysis.extraPaths": [
			"Apps/maya/scripts", // <- Should work but doesn't
			//"../../../../../game_p9/bin/tools/Apps/maya/scripts", // <- Works
		],
	}
}
```

## Repro Steps

1. In a Python module in the above workspace, I try to import a Python module in the maya/Scripts directory.

## Expected behavior

Pylance treats relative paths in python.analysis.extrapaths as being relative to the .code-workspace file for the current workspace.
Using those paths it can find the module and is able to provide code completion.

## Actual behavior

Pylance cannot find any modules in maya/Scripts unless I change the extraPaths value to the second one shown, where the path is relative to python.defaultInterpreterPath (or maybe something else similar), instead of being relative to the workspace root directory.

Absolute paths also work.

The VS Code Python documentation example ["path 1","path 2"] is not clear how paths should be entered but I have gathered from comments on GitHub issues that paths should be relative to the workspace root. I assume the workspace root is the location of the .code-workspace file.

That documentation appears to say that VS Code variables such as ${workspaceFolder} should work, but from GitHub comments from Pylance devs I suspect they actually aren't supported. None that I've tried work for me.

I have also tried to use windows environment variables like ${env:MY_ENV_VAR} but these do not appear to work either.

Maybe this is not the documentation I should be using, though?
https://code.visualstudio.com/docs/python/settings-reference

The Pylance-specific readme at https://github.com/microsoft/pylance-release/blob/main/README.md
doesn't address these issues so I'm not sure what the behavior should be.

**Comment by Boon-in-Oz** [comment]
Thanks.

How do I know what the workspace folder is? This is a multi-root workspace. Is it one of the root folders? Or is there a separate setting for it? In my experience ${workspaceFolder} varies unpredictably, which is why I used ${workspaceFolder:TechTools} in my example above.

**Comment by rchiodo** [comment]
I believe that's a bug then. We don't seem to support the multi root case and referencing the folder roots. 

You can't use this at the moment:

```
"${workspaceFolder:TechTools}/Scripts"
```

However, this should work:

```
"${workspaceFolder}/Scripts"
```

Under the covers, pylance is expanding the environment variables in the settings based on what it knows about the workspace.  With that setting, it will add the `Scripts` subfolder to the extraPaths. 

Meaning every workspace folder will have a 'Scripts' folder added to the extraPaths, it just won't find anything in the other two folders.

**Comment by Boon-in-Oz** [comment]
Thanks. Adding the ${workspaceFolder:root} expansion sounds like it'll work well for me.

Just some more clarity on the current behavior:

I think what I'm seeing is that, in python.analysis.extraPaths: 
`"${workspaceFolder}/Scripts"`
only works for the current workspace folder. That is, if I'm typing in a module inside the "maya" folder, autocomplete works for modules and packages imported relative to maya/Scripts, but if I'm typing in a module in "TechTools", maya/Scripts is no longer available.

Whereas if I just add 
`"Scripts"`
I get the same result no matter where I'm typing: I can import relative to ../../../game/bin/teams/techart/packages/Scripts.

I'm not sure why ../../../game/bin/teams/techart/packages is regarded as the cwd for extraPaths. Perhaps because it's the first entry in "folders" alphabetically?

**Comment by Boon-in-Oz** [comment]
OK one more thing. I'm seeing inconsistent interpretation of the ${workspaceFolder:folder} variable from VS Code. I know this isn't in your code but it might affect your implementation of this fix.

Recall this part of my code-workspace file:
```
		{
			"name": "Techart Packages",
			"path": "../../../game/bin/teams/techart/packages"
		},
```
If I set `"python.defaultInterpreterPath": "${workspaceFolder:Techart Packages}/../python3.7.7/python.exe"`, then create a task in my workspace like this:
```
			{
				"label": "echo python default interpreter",
				"type": "shell",
				"command": "echo python.defaultInterpreterPath: ${config:python.defaultInterpreterPath}"
			},
```
it works. But the Python extension complains "An invalid interpreter is selected, ..." and several of "An invalid interpreter is selected for workspace..."

Whereas if I set it like this: `"python.defaultInterpreterPath": "${workspaceFolder:packages}/../python3.7.7/python.exe"`, the Python extension no longer complains, but the task fails with `Variable ${workspaceFolder:packages} can not be resolved. No such fo..."

**Comment by hangie** [comment]
I'm having the same kind of problem, however if I create a similar Task and echo the content of: ${config:python.analysis.extraPaths} it is showing the correct directories, however pylance doesn't seem to be using them to resolve my packages.

**Comment by nrs-mparker** [comment]
I'm experiencing this as well.  Trying to do a setup for an Unreal project like:
```json
{
	"folders": [
		{
			"name": "MyGame",
			"path": "../.."
		}
		{
			"name": "Movie Render Pipeline",
			"path": "../../../../../Engine/Plugins/MovieScene/MovieRenderPipeline/Content/Python"
		}
	],
	"settings": {
		"python.defaultInterpreterPath": "${workspaceFolder:MyGame}/../../../Engine/Binaries/ThirdParty/Python3/Win64/python.exe",
		"python.analysis.extraPaths": [
			"${workspaceFolder:MyGame}/../../Intermediate/PythonStub"
		]
	}
}
```

The ability for defaultInterpreterPath to reference named workspace folders was fixed a couple weeks ago, but now I'm running into that issue here.  There are more roots in the .code-workspace than this, they vary in how many directory levels deep they are.  All of them need the stub from the same directory.  It can't be an absolute path because this will be shared between developers who have the same structure but possibly differing absolute folders.  With the current support for `${workspaceFolder}` and relative paths, I can make a single one work (or multiple, if they happen to be the same level of nested folders), but I can't make all of them work like they would if I could use `${workspaceFolder:MyGame}`.  This could also be solved if VS Code had some kind of `${workspaceFile}` that pointed to where the .code-workspace file was located, but they have indicated they do not intend to add that.

**Comment by Seairth** [comment]
> I'm having the same kind of problem, however if I create a similar Task and echo the content of: ${config:python.analysis.extraPaths} it is showing the correct directories, however pylance doesn't seem to be using them to resolve my packages.
> 
> If I take the results of the expansion in the echo command and paste this back in replacing the use of "${workspaceFolder:my-module}" it can expand my modules just fine.

I ended up here after doing the same exact test.

Digging a bit further, it looks like the variable expansion isn't automatically handled by vscode, but is the responsibility of the extension.  In the case of the _echo_ task, VS Code is doing the work.  And, in the case of `extraPaths`, it looks like it's pyright that's doing the actual work (see https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/server.ts#L169). 
 And based on comments in the pyright code, it seems that VSCode doesn't provide a convenient API to handle expansion of all the "standard" variables. So, to get this working, I guess it requires an update to pyright? :thinking:

**Comment by rchiodo** [comment]
> So, to get this working, I guess it requires an update to pyright? 🤔

That is correct. You could submit a PR if you were interested as this resides mostly in Pyright. Otherwise that's probably how we'll fix it, add logic to handle the ':' in the expansion.

I believe the expansion actually happens here:
https://github.com/microsoft/pyright/blob/9fb26eb7a3c8b2dc68ba011981e9966069972c04/packages/pyright-internal/src/common/envVarUtils.ts#L26

That function would likely have to be updated to somehow get all of the open workspaces and their names. Which is passed in the LSP messages to Pyright. Meaning this should be possible to fix, we just haven't gotten to it yet.

**Comment by Seairth** [comment]
The change has been merged into pyright (https://github.com/microsoft/pyright/pull/7138).

**Comment by rchiodo** [comment]
The latest Pylance release has this fix in it.

**Comment by samskiter** [comment]
While I have just found this issue - I am not convinced it is actually fixed. I have a multi-root workspace and can't get autocomplete/import search to work. I've updated the pylance extension to v2024.2.105 (pre-release) but can't get intellisense working to auto import or code complete from the other folder that is included in the workspace. manually typing out the import shows all the right intellisense colours
